src/twisted/mail/test/test_mail.py

```diff
@@ -247,25 +235,99 @@ def _cbFinalName(self, result):
         self.assertTrue(self.f.closed)
         self.assertFalse(os.path.exists(self.name))
 
-    @skipIf(sys.version_info >= (3,), "not ported to Python 3")
     def testContents(self):
-        contents = "first line\nsecond line\nthird line\n"
+        contents = b"first line\nsecond line\nthird line\n"
         for line in contents.splitlines():
             self.fp.lineReceived(line)
         self.fp.eomReceived()
-        with open(self.final) as f:
+        with open(self.final, "rb") as f:
             self.assertEqual(f.read(), contents)
 
-    @skipIf(sys.version_info >= (3,), "not ported to Python 3")
     def testInterrupted(self):
-        contents = "first line\nsecond line\n"
+        contents = b"first line\nsecond line\n"
         for line in contents.splitlines():
             self.fp.lineReceived(line)
         self.fp.connectionLost()
         self.assertFalse(os.path.exists(self.name))
         self.assertFalse(os.path.exists(self.final))
 
 
+@skipIf(platformType != "posix", "twisted.mail only works on posix")
+class MaildirMessageTests(TestCase):
+    """
+    Tests for the file creating by the L{mail.maildir.MaildirMessage}.
+    """
+
+    def setUp(self):
+        """
+        Create and open a temporary file.
+        """
+        self.name = self.mktemp()
+        self.final = self.mktemp()
+        self.address = b"user@example.com"
+        self.f = open(self.name, "wb")
+        self.addCleanup(self.f.close)
+        self.fp = mail.maildir.MaildirMessage(
+            self.address, self.f, self.name, self.final
+        )
+
+    def _finalName(self):
+        """
+        Search for the final file path.
+
+        @rtype: L{str}
+        @return: Final file path.
+        """
+        return glob.glob(f"{self.final},S=[0-9]*")[0]
+
+    def test_finalName(self):
+        """
+        Send the EOM to the message and add the callback for testing final name.
+        """
+        self.fp.eomReceived().addCallback(self._cbFinalName)
```

Comment:
You must either `return` the `Deferred` (so that the test waits on it) or assert that it resolves synchronously like:

```py
self.successResultOf(self.fp.eomReceived().addCallback(self._cbFinalName))
```

You could also write that like:

```py
result = self.successResultOf(self.fp.eomReceived())
self._cbFinalName(result)
```

If you prefer this latter style I'd also rename `_cbFinalName` to `assertFinalName` as it's no longer functioning as a callback.

Reply:
Just a minor comment. Maybe is easier to rewrite the test with `defer.inlineCallback` ... 

Or maybe trial already support async syntax. I am not familiar with py3 syntax, but `defer.inlineCallbacks` makes the code so much easier to read.